File: lang/csharp/Eclipse.eCAL.Core.Test/ConfigTest.cs
/* ========================= eCAL LICENSE =================================
 *
 * Copyright (C) 2016 - 2025 Continental Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * ========================= eCAL LICENSE =================================
*/

/**
 * @file ConfigTest.cs
 * @brief Tests for type wrapper and initializing.
 */

using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using System.Threading;
using Eclipse.eCAL.Core;
using Eclipse.eCAL.Core.Config;
using Microsoft.VisualStudio.TestTools.UnitTesting;

[TestClass]
public class ConfigTest
{
  private const int TcsWaitTime = 2500;

  [TestMethod]
  public void TestConfigPassing()
  {
    // Create config with every value set differently than the default values
    var config = new Eclipse.eCAL.Core.Config.Configuration();

    // Communication mode
    config.CommunicationMode = eCommunicationMode.Network;

    // Registration
    config.Registration.RegistrationRefresh = 500;
    config.Registration.RegistrationTimeout = 2000;
    config.Registration.Loopback = false;
    config.Registration.ShmTransportDomain = "shm_transport_domain";
    config.Registration.Local.TransportType = 0; // eCAL::Registration::Local::eTransportType::shm
    config.Registration.Local.SHM.Domain = "ecal_don";
    config.Registration.Local.SHM.QueueSize = 2048;
    config.Registration.Local.UDP.Port = 15000;
    config.Registration.Network.TransportType = 0; // eCAL::Registration::Network::eTransportType::udp
    config.Registration.Network.UDP.Port = 16000;

    // Transport Layer
    config.TransportLayer.Udp.ConfigVersion = TypesUdpConfigVersion.V1;
    config.TransportLayer.Udp.Port = 17000;
    config.TransportLayer.Udp.Mask = "255.254.254.242";
    config.TransportLayer.Udp.SendBuffer = 6242880;
    config.TransportLayer.Udp.ReceiveBuffer = 6242881;
    config.TransportLayer.Udp.JoinAllInterfaces = true;
    config.TransportLayer.Udp.NpcapEnabled = true;
    config.TransportLayer.Udp.Local.Group = "129.255.255.254";
    config.TransportLayer.Udp.Local.Ttl = 7;
    config.TransportLayer.Udp.Network.Group = "238.1.2.3";
    config.TransportLayer.Udp.Network.Ttl = 8;
    config.TransportLayer.Tcp.NumberExecutorReader = 9;
    config.TransportLayer.Tcp.NumberExecutorWriter = 10;
    config.TransportLayer.Tcp.MaxReconnections = 11;

    // Publisher
    config.Publisher.Layer.SHM.Enable = false;
    config.Publisher.Layer.SHM.ZeroCopyMode = true;
    config.Publisher.Layer.SHM.AcknowledgeTimeoutMs = 12;
    config.Publisher.Layer.SHM.MemfileBufferCount = 13;
    config.Publisher.Layer.SHM.MemfileMinSizeBytes = 8192;
    config.Publisher.Layer.SHM.MemfileReservePercent = 14;
    config.Publisher.Layer.UDP.Enable = false;
    config.Publisher.Layer.TCP.Enable = false;
    config.Publisher.LayerPriorityLocal.Clear();
    config.Publisher.LayerPriorityLocal.AddRange(new eTransportLayerType[] { eTransportLayerType.Tcp, eTransportLayerType.Shm, eTransportLayerType.UdpMc });
    config.Publisher.LayerPriorityRemote.Clear();
    config.Publisher.LayerPriorityRemote.AddRange(new eTransportLayerType[] { eTransportLayerType.Tcp, eTransportLayerType.UdpMc });

    // Subscriber
    config.Subscriber.Layer.SHM.Enable = false;
    config.Subscriber.Layer.UDP.Enable = false;
    config.Subscriber.Layer.TCP.Enable = true;
    config.Subscriber.DropOutOfOrderMessages = false;

    // TimeSync
    config.TimeSync.TimeSyncModuleReplay = "my_replay";
    config.TimeSync.TimeSyncModuleRT = "my_rt";

    // Application
    config.Application.Startup.TerminalEmulator = "term_emulator";
    config.Application.Sys.FilterExcl = "filter_excl";

    // Logging
    config.Logging.Provider.Console.Enable = false;
    config.Logging.Provider.Console.LogLevel = 1; // log_level_debug1
    config.Logging.Provider.File.Enable = true;
    config.Logging.Provider.File.LogLevel = 2 | 3; // log_level_debug2 | log_level_debug3
    config.Logging.Provider.UDP.Enable = false;
    config.Logging.Provider.UDP.LogLevel = 4; // log_level_debug4
    config.Logging.Provider.FileConfig.Path = "file_config_path";
    config.Logging.Provider.UDPConfig.Port = 18000;
    config.Logging.Receiver.Enable = true;
    config.Logging.Receiver.UDPConfig.Port = 19000;

    Core.Initialize(config, "TestConfigPassing", Init.Flags.Default);

    var ecalConfig = Config.GetConfiguration();

    // Communication mode
    Assert.AreEqual(config.CommunicationMode, ecalConfig.CommunicationMode, "CommunicationMode mismatch");

    // Registration
    Assert.AreEqual(config.Registration.RegistrationRefresh, ecalConfig.Registration.RegistrationRefresh, "RegistrationRefresh mismatch");
    Assert.AreEqual(config.Registration.RegistrationTimeout, ecalConfig.Registration.RegistrationTimeout, "RegistrationTimeout mismatch");
    Assert.AreEqual(config.Registration.Loopback, ecalConfig.Registration.Loopback, "Loopback mismatch");
    Assert.AreEqual(config.Registration.ShmTransportDomain, ecalConfig.Registration.ShmTransportDomain, "ShmTransportDomain mismatch");
    Assert.AreEqual(config.Registration.Local.TransportType, ecalConfig.Registration.Local.TransportType, "Local.TransportType mismatch");
    Assert.AreEqual(config.Registration.Local.SHM.Domain, ecalConfig.Registration.Local.SHM.Domain, "Local.SHM.Domain mismatch");
    Assert.AreEqual(config.Registration.Local.SHM.QueueSize, ecalConfig.Registration.Local.SHM.QueueSize, "Local.SHM.QueueSize mismatch");
    Assert.AreEqual(config.Registration.Local.UDP.Port, ecalConfig.Registration.Local.UDP.Port, "Local.UDP.Port mismatch");
    Assert.AreEqual(config.Registration.Network.TransportType, ecalConfig.Registration.Network.TransportType, "Network.TransportType mismatch");
    Assert.AreEqual(config.Registration.Network.UDP.Port, ecalConfig.Registration.Network.UDP.Port, "Network.UDP.Port mismatch");

    // Transport Layer
    Assert.AreEqual(config.TransportLayer.Udp.ConfigVersion, ecalConfig.TransportLayer.Udp.ConfigVersion, "Udp.ConfigVersion mismatch");
    Assert.AreEqual(config.TransportLayer.Udp.Port, ecalConfig.TransportLayer.Udp.Port, "Udp.Port mismatch");
    Assert.AreEqual(config.TransportLayer.Udp.Mask, ecalConfig.TransportLayer.Udp.Mask, "Udp.Mask mismatch");
    Assert.AreEqual(config.TransportLayer.Udp.SendBuffer, ecalConfig.TransportLayer.Udp.SendBuffer, "Udp.SendBuffer mismatch");
    Assert.AreEqual(config.TransportLayer.Udp.ReceiveBuffer, ecalConfig.TransportLayer.Udp.ReceiveBuffer, "Udp.ReceiveBuffer mismatch");
    Assert.AreEqual(config.TransportLayer.Udp.JoinAllInterfaces, ecalConfig.TransportLayer.Udp.JoinAllInterfaces, "Udp.JoinAllInterfaces mismatch");
    Assert.AreEqual(config.TransportLayer.Udp.NpcapEnabled, ecalConfig.TransportLayer.Udp.NpcapEnabled, "Udp.NpcapEnabled mismatch");
    Assert.AreEqual(config.TransportLayer.Udp.Local.Group, ecalConfig.TransportLayer.Udp.Local.Group, "Udp.Local.Group mismatch");
    Assert.AreEqual(config.TransportLayer.Udp.Local.Ttl, ecalConfig.TransportLayer.Udp.Local.Ttl, "Udp.Local.Ttl mismatch");
    Assert.AreEqual(config.TransportLayer.Udp.Network.Group, ecalConfig.TransportLayer.Udp.Network.Group, "Udp.Network.Group mismatch");
    Assert.AreEqual(config.TransportLayer.Udp.Network.Ttl, ecalConfig.TransportLayer.Udp.Network.Ttl, "Udp.Network.Ttl mismatch");
    Assert.AreEqual(config.TransportLayer.Tcp.NumberExecutorReader, ecalConfig.TransportLayer.Tcp.NumberExecutorReader, "Tcp.NumberExecutorReader mismatch");
    Assert.AreEqual(config.TransportLayer.Tcp.NumberExecutorWriter, ecalConfig.TransportLayer.Tcp.NumberExecutorWriter, "Tcp.NumberExecutorWriter mismatch");
    Assert.AreEqual(config.TransportLayer.Tcp.MaxReconnections, ecalConfig.TransportLayer.Tcp.MaxReconnections, "Tcp.MaxReconnections mismatch");

    // Publisher
    Assert.AreEqual(config.Publisher.Layer.SHM.Enable, ecalConfig.Publisher.Layer.SHM.Enable, "Publisher.Layer.SHM.Enable mismatch");
    Assert.AreEqual(config.Publisher.Layer.SHM.ZeroCopyMode, ecalConfig.Publisher.Layer.SHM.ZeroCopyMode, "Publisher.Layer.SHM.ZeroCopyMode mismatch");
    Assert.AreEqual(config.Publisher.Layer.SHM.AcknowledgeTimeoutMs, ecalConfig.Publisher.Layer.SHM.AcknowledgeTimeoutMs, "Publisher.Layer.SHM.AcknowledgeTimeoutMs mismatch");
    Assert.AreEqual(config.Publisher.Layer.SHM.MemfileBufferCount, ecalConfig.Publisher.Layer.SHM.MemfileBufferCount, "Publisher.Layer.SHM.MemfileBufferCount mismatch");
    Assert.AreEqual(config.Publisher.Layer.SHM.MemfileMinSizeBytes, ecalConfig.Publisher.Layer.SHM.MemfileMinSizeBytes, "Publisher.Layer.SHM.MemfileMinSizeBytes mismatch");
    Assert.AreEqual(config.Publisher.Layer.SHM.MemfileReservePercent, ecalConfig.Publisher.Layer.SHM.MemfileReservePercent, "Publisher.Layer.SHM.MemfileReservePercent mismatch");
    Assert.AreEqual(config.Publisher.Layer.UDP.Enable, ecalConfig.Publisher.Layer.UDP.Enable, "Publisher.Layer.UDP.Enable mismatch");
    Assert.AreEqual(config.Publisher.Layer.TCP.Enable, ecalConfig.Publisher.Layer.TCP.Enable, "Publisher.Layer.TCP.Enable mismatch");
    CollectionAssert.AreEqual(config.Publisher.LayerPriorityLocal, ecalConfig.Publisher.LayerPriorityLocal, "Publisher.LayerPriorityLocal mismatch");
    CollectionAssert.AreEqual(config.Publisher.LayerPriorityRemote, ecalConfig.Publisher.LayerPriorityRemote, "Publisher.LayerPriorityRemote mismatch");

    // Subscriber
    Assert.AreEqual(config.Subscriber.Layer.SHM.Enable, ecalConfig.Subscriber.Layer.SHM.Enable, "Subscriber.Layer.SHM.Enable mismatch");
    Assert.AreEqual(config.Subscriber.Layer.UDP.Enable, ecalConfig.Subscriber.Layer.UDP.Enable, "Subscriber.Layer.UDP.Enable mismatch");
    Assert.AreEqual(config.Subscriber.Layer.TCP.Enable, ecalConfig.Subscriber.Layer.TCP.Enable, "Subscriber.Layer.TCP.Enable mismatch");
    Assert.AreEqual(config.Subscriber.DropOutOfOrderMessages, ecalConfig.Subscriber.DropOutOfOrderMessages, "Subscriber.DropOutOfOrderMessages mismatch");

    // TimeSync
    Assert.AreEqual(config.TimeSync.TimeSyncModuleReplay, ecalConfig.TimeSync.TimeSyncModuleReplay, "TimeSync.TimeSyncModuleReplay mismatch");
    Assert.AreEqual(config.TimeSync.TimeSyncModuleRT, ecalConfig.TimeSync.TimeSyncModuleRT, "TimeSync.TimeSyncModuleRT mismatch");

    // Application
    Assert.AreEqual(config.Application.Startup.TerminalEmulator, ecalConfig.Application.Startup.TerminalEmulator, "Application.Startup.TerminalEmulator mismatch");
    Assert.AreEqual(config.Application.Sys.FilterExcl, ecalConfig.Application.Sys.FilterExcl, "Application.Sys.FilterExcl mismatch");

    // Logging
    Assert.AreEqual(config.Logging.Provider.Console.Enable, ecalConfig.Logging.Provider.Console.Enable, "Logging.Provider.Console.Enable mismatch");
    Assert.AreEqual(config.Logging.Provider.Console.LogLevel, ecalConfig.Logging.Provider.Console.LogLevel, "Logging.Provider.Console.LogLevel mismatch");
    Assert.AreEqual(config.Logging.Provider.File.Enable, ecalConfig.Logging.Provider.File.Enable, "Logging.Provider.File.Enable mismatch");
    Assert.AreEqual(config.Logging.Provider.File.LogLevel, ecalConfig.Logging.Provider.File.LogLevel, "Logging.Provider.File.LogLevel mismatch");
    Assert.AreEqual(config.Logging.Provider.UDP.Enable, ecalConfig.Logging.Provider.UDP.Enable, "Logging.Provider.UDP.Enable mismatch");
    Assert.AreEqual(config.Logging.Provider.UDP.LogLevel, ecalConfig.Logging.Provider.UDP.LogLevel, "Logging.Provider.UDP.LogLevel mismatch");
    Assert.AreEqual(config.Logging.Provider.FileConfig.Path, ecalConfig.Logging.Provider.FileConfig.Path, "Logging.Provider.FileConfig.Path mismatch");
    Assert.AreEqual(config.Logging.Provider.UDPConfig.Port, ecalConfig.Logging.Provider.UDPConfig.Port, "Logging.Provider.UDPConfig.Port mismatch");
    Assert.AreEqual(config.Logging.Receiver.Enable, ecalConfig.Logging.Receiver.Enable, "Logging.Receiver.Enable mismatch");
    Assert.AreEqual(config.Logging.Receiver.UDPConfig.Port, ecalConfig.Logging.Receiver.UDPConfig.Port, "Logging.Receiver.UDPConfig.Port mismatch");

    Core.Terminate();
  }

  [TestMethod]
  public void TestPubSubConfig()
  {
    Core.Initialize("TestPubSubConfig");

    // Setting up SHM publisher and subscriber
    var publisherConfiguration = Config.GetPublisherConfiguration();
    publisherConfiguration.Layer.SHM.Enable = true;
    publisherConfiguration.Layer.UDP.Enable = false;
    publisherConfiguration.Layer.TCP.Enable = false;

    var subscriberConfiguration = Config.GetSubscriberConfiguration();
    subscriberConfiguration.Layer.SHM.Enable = true;
    subscriberConfiguration.Layer.UDP.Enable = false;
    subscriberConfiguration.Layer.TCP.Enable = false;

    Publisher publisherSHM = new Publisher("PubSubConfig", null, null, publisherConfiguration);
    Subscriber subscriberSHM = new Subscriber("PubSubConfig", null, null, subscriberConfiguration);

    var tcsSHM = new TaskCompletionSource<(string message, long timestamp)>();

    subscriberSHM.SetReceiveCallback((publisherId, dtInfo, data) =>
    {
      string received = Encoding.UTF8.GetString(data.Buffer);
      tcsSHM.TrySetResult((received, data.SendTimestamp));
    });

    // Setting up UDP publisher and subscriber
    publisherConfiguration.Layer.SHM.Enable = false;
    publisherConfiguration.Layer.UDP.Enable = true;
    publisherConfiguration.Layer.TCP.Enable = false;

    subscriberConfiguration.Layer.SHM.Enable = false;
    subscriberConfiguration.Layer.UDP.Enable = true;
    subscriberConfiguration.Layer.TCP.Enable = false;

    Publisher publisherUDP = new Publisher("PubSubConfig", null, null, publisherConfiguration);
    Subscriber subscriberUDP = new Subscriber("PubSubConfig", null, null, subscriberConfiguration);

    var tcsUDP = new TaskCompletionSource<(string message, long timestamp)>();

    subscriberUDP.SetReceiveCallback((publisherId, dtInfo, data) =>
    {
      string received = Encoding.UTF8.GetString(data.Buffer);
      tcsUDP.TrySetResult((received, data.SendTimestamp));
    });

    // Setting up TCP publisher and subscriber
    publisherConfiguration.Layer.SHM.Enable = false;
    publisherConfiguration.Layer.UDP.Enable = false;
    publisherConfiguration.Layer.TCP.Enable = true;

    subscriberConfiguration.Layer.SHM.Enable = false;
    subscriberConfiguration.Layer.UDP.Enable = false;
    subscriberConfiguration.Layer.TCP.Enable = true;

    Publisher publisherTCP = new Publisher("PubSubConfig", null, null, publisherConfiguration);
    Subscriber subscriberTCP = new Subscriber("PubSubConfig", null, null, subscriberConfiguration);

    var tcsTCP = new TaskCompletionSource<(string message, long timestamp)>();

    subscriberTCP.SetReceiveCallback((publisherId, dtInfo, data) =>
    {
      string received = Encoding.UTF8.GetString(data.Buffer);
      tcsTCP.TrySetResult((received, data.SendTimestamp));
    });

    string messageSHM = "Sending via SHM";
    string messageUDP = "Sending via UDP";
    string messageTCP = "Sending via TCP";

    // Sleep a moment for connections to be established
    Thread.Sleep(2000);

    publisherSHM.Send(Encoding.UTF8.GetBytes(messageSHM));

    if (!tcsSHM.Task.Wait(TcsWaitTime))
    {
      Assert.Fail("Timeout waiting for SHM message.");
    }

    var resultSHM = tcsSHM.Task.Result;

    Assert.IsTrue(tcsSHM.Task.IsCompleted, "SHM Task not completed.");
    Assert.AreEqual(messageSHM, resultSHM.message, "SHM Received message mismatch");

    // Check that UDP and TCP was not called
    Assert.IsFalse(tcsUDP.Task.IsCompleted, "UDP callback should not have been called.");
    Assert.IsFalse(tcsTCP.Task.IsCompleted, "TCP callback should not have been called.");

    publisherUDP.Send(Encoding.UTF8.GetBytes(messageUDP));

    if (!tcsUDP.Task.Wait(TcsWaitTime))
    {
      Assert.Fail("Timeout waiting for UDP message.");
    }

    var resultUDP = tcsUDP.Task.Result;
    Assert.IsTrue(tcsUDP.Task.IsCompleted, "UDP Task not completed.");
    Assert.AreEqual(messageUDP, resultUDP.message, "UDP received message mismatch.");

    // Check that TCP was not called
    Assert.IsFalse(tcsTCP.Task.IsCompleted, "TCP callback should not have been called.");

    publisherTCP.Send(Encoding.UTF8.GetBytes(messageTCP));

    if (!tcsTCP.Task.Wait(TcsWaitTime))
    {
      Assert.Fail("Timeout waiting for TCP message.");
    }

    var resultTCP = tcsTCP.Task.Result;
    Assert.IsTrue(tcsTCP.Task.IsCompleted, "TCP Task not completed.");
    Assert.AreEqual(messageTCP, resultTCP.message, "TCP received message mismatch.");

    Core.Terminate();
  }

  [TestMethod]
  public void TestConfigConvenienceGetter()
  {
    var ecalConfig = Config.GetConfiguration();

    // Now check all convenience getters against the configuration object
    Assert.AreEqual(ecalConfig.Registration.RegistrationTimeout, Config.GetRegistrationTimeoutMs(), "RegistrationTimeout");
    Assert.AreEqual(ecalConfig.Registration.RegistrationRefresh, Config.GetRegistrationRefreshMs(), "RegistrationRefresh");
    Assert.AreEqual(ecalConfig.CommunicationMode == eCommunicationMode.Local, Config.IsNetworkEnabled(), "IsNetworkEnabled");
    Assert.AreEqual(ecalConfig.TimeSync.TimeSyncModuleRT, Config.GetTimesyncModuleName(), "TimeSyncModuleRT");
    Assert.AreEqual(ecalConfig.TimeSync.TimeSyncModuleReplay, Config.GetTimesyncModuleReplay(), "TimeSyncModuleReplay");
    Assert.AreEqual(ecalConfig.Application.Startup.TerminalEmulator, Config.GetTerminalEmulatorCommand(), "TerminalEmulator");
    Assert.AreEqual(ecalConfig.Application.Sys.FilterExcl, Config.GetEcalSysFilterExcludeList(), "FilterExcl");
    Assert.AreEqual(ecalConfig.Subscriber.DropOutOfOrderMessages, Config.GetDropOutOfOrderMessages(), "DropOutOfOrderMessages");
    Assert.AreEqual(ecalConfig.Registration.Local.SHM.QueueSize, Config.GetShmMonitoringQueueSize(), "ShmMonitoringQueueSize");
    Assert.AreEqual(ecalConfig.Registration.Local.SHM.Domain, Config.GetShmMonitoringDomain(), "ShmMonitoringDomain");
    Assert.AreEqual(ecalConfig.Registration.Local.TransportType == 0 && ecalConfig.CommunicationMode == 0, Config.IsShmRegistrationEnabled(), "IsShmRegistrationEnabled");
    Assert.AreEqual(ecalConfig.TransportLayer.Udp.ConfigVersion, Config.GetUdpMulticastConfigVersion(), "UdpMulticastConfigVersion");
    Assert.AreEqual(ecalConfig.TransportLayer.Udp.Network.Group, Config.GetUdpMulticastGroup(), "UdpMulticastGroup");
    Assert.AreEqual(ecalConfig.TransportLayer.Udp.Mask, Config.GetUdpMulticastMask(), "UdpMulticastMask");
    Assert.AreEqual(ecalConfig.TransportLayer.Udp.Port, Config.GetUdpMulticastPort(), "UdpMulticastPort");
    Assert.AreEqual(ecalConfig.TransportLayer.Udp.Network.Ttl, Config.GetUdpMulticastTtl(), "UdpMulticastTtl");
    Assert.AreEqual(ecalConfig.TransportLayer.Udp.SendBuffer, Config.GetUdpMulticastSndBufSizeBytes(), "UdpMulticastSndBufSizeBytes");
    Assert.AreEqual(ecalConfig.TransportLayer.Udp.ReceiveBuffer, Config.GetUdpMulticastRcvBufSizeBytes(), "UdpMulticastRcvBufSizeBytes");
    Assert.AreEqual(ecalConfig.TransportLayer.Udp.JoinAllInterfaces, Config.IsUdpMulticastJoinAllIfEnabled(), "IsUdpMulticastJoinAllIfEnabled");
    Assert.AreEqual(ecalConfig.Subscriber.Layer.UDP.Enable, Config.IsUdpMulticastRecEnabled(), "IsUdpMulticastRecEnabled");
    Assert.AreEqual(ecalConfig.Subscriber.Layer.SHM.Enable, Config.IsShmRecEnabled(), "IsShmRecEnabled");
    Assert.AreEqual(ecalConfig.Subscriber.Layer.TCP.Enable, Config.IsTcpRecEnabled(), "IsTcpRecEnabled");
    Assert.AreEqual(ecalConfig.TransportLayer.Udp.NpcapEnabled, Config.IsNpcapEnabled(), "IsNpcapEnabled");
    Assert.AreEqual(ecalConfig.TransportLayer.Tcp.NumberExecutorReader, Config.GetTcpPubsubReaderThreadpoolSize(), "GetTcpPubsubReaderThreadpoolSize");
    Assert.AreEqual(ecalConfig.TransportLayer.Tcp.NumberExecutorWriter, Config.GetTcpPubsubWriterThreadpoolSize(), "GetTcpPubsubWriterThreadpoolSize");
    Assert.AreEqual(ecalConfig.TransportLayer.Tcp.MaxReconnections, Config.GetTcpPubsubMaxReconnectionAttemps(), "GetTcpPubsubMaxReconnectionAttemps");
    Assert.AreEqual(ecalConfig.Registration.ShmTransportDomain, Config.GetShmTransportDomain(), "GetShmTransportDomain");
  }
}

File: lang/c/tests/config_test/src/config_test.cpp
/* ========================= eCAL LICENSE =================================
 *
 * Copyright (C) 2016 - 2025 Continental Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * ========================= eCAL LICENSE =================================
*/

#include <ecal_c/ecal.h>
#include <ecal/ecal.h>
#include <gtest/gtest.h>

class config_test_c : public ::testing::Test 
{
    protected:
        eCAL_Configuration* configuration0;

        void SetUp() override 
        {
            configuration0 = eCAL_Configuration_New();
            eCAL_Configuration_InitFromConfig(configuration0);
            eCAL_Initialize(NULL, NULL, configuration0);
        }

        void TearDown() override 
        {
            eCAL_Finalize();
            eCAL_Configuration_Delete(configuration0);
        }
};

TEST_F(config_test_c, GetConfiguration)
{
    EXPECT_NE(nullptr, eCAL_GetConfiguration());
    EXPECT_NE(nullptr, eCAL_GetSubscriberConfiguration());
    EXPECT_NE(nullptr, eCAL_GetPublisherConfiguration());
}

TEST_F(config_test_c, Network)
{
    EXPECT_EQ(configuration0->communication_mode, eCAL_Config_IsNetworkEnabled());
}

TEST_F(config_test_c, Registration) 
{
    EXPECT_EQ(configuration0->registration.registration_timeout, eCAL_Config_GetRegistrationTimeoutMs());
    EXPECT_EQ(configuration0->registration.registration_timeout, eCAL_GetConfiguration()->registration.registration_timeout);
    EXPECT_EQ(configuration0->registration.registration_refresh, eCAL_Config_GetRegistrationRefreshMs());
    EXPECT_EQ(configuration0->registration.registration_refresh, eCAL_GetConfiguration()->registration.registration_refresh);
    //EXPECT_EQ(configuration0->registration.local.transport_type , eCAL_Config_IsShmRegistrationEnabled());
    EXPECT_STREQ(configuration0->registration.shm_transport_domain, eCAL_Config_GetShmTransportDomain());
    EXPECT_STREQ(configuration0->registration.shm_transport_domain, eCAL_GetConfiguration()->registration.shm_transport_domain);
    EXPECT_EQ(configuration0->registration.local.shm.queue_size, eCAL_Config_GetShmMonitoringQueueSize());
    EXPECT_EQ(configuration0->registration.local.shm.queue_size, eCAL_GetConfiguration()->registration.local.shm.queue_size);
    EXPECT_STREQ(configuration0->registration.local.shm.domain, eCAL_Config_GetShmMonitoringDomain());
    EXPECT_STREQ(configuration0->registration.local.shm.domain, eCAL_GetConfiguration()->registration.local.shm.domain);

    EXPECT_EQ(configuration0->registration.local.udp.port, eCAL_GetConfiguration()->registration.local.udp.port);
    EXPECT_EQ(configuration0->registration.loopback, eCAL_GetConfiguration()->registration.loopback);
    EXPECT_EQ(configuration0->registration.network.transport_type, eCAL_GetConfiguration()->registration.network.transport_type);
    EXPECT_EQ(configuration0->registration.network.udp.port, eCAL_GetConfiguration()->registration.network.udp.port);
    EXPECT_STREQ(configuration0->registration.shm_transport_domain, eCAL_GetConfiguration()->registration.shm_transport_domain);
}

TEST_F(config_test_c, TransportLayer) 
{
    EXPECT_EQ(configuration0->transport_layer.udp.config_version, eCAL_Config_GetUdpMulticastConfigVersion());
    EXPECT_EQ(configuration0->transport_layer.udp.config_version, eCAL_GetConfiguration()->transport_layer.udp.config_version);
    EXPECT_STREQ(configuration0->transport_layer.udp.network.group, eCAL_Config_GetUdpMulticastGroup());
    EXPECT_STREQ(configuration0->transport_layer.udp.network.group, eCAL_GetConfiguration()->transport_layer.udp.network.group);
    EXPECT_STREQ(configuration0->transport_layer.udp.mask, eCAL_Config_GetUdpMulticastMask());
    EXPECT_STREQ(configuration0->transport_layer.udp.mask, eCAL_GetConfiguration()->transport_layer.udp.mask);
    EXPECT_EQ(configuration0->transport_layer.udp.port, eCAL_Config_GetUdpMulticastPort());
    EXPECT_EQ(configuration0->transport_layer.udp.port, eCAL_GetConfiguration()->transport_layer.udp.port);
    EXPECT_EQ(configuration0->transport_layer.udp.network.ttl, eCAL_Config_GetUdpMulticastTtl());
    EXPECT_EQ(configuration0->transport_layer.udp.network.ttl, eCAL_GetConfiguration()->transport_layer.udp.network.ttl);
    EXPECT_EQ(configuration0->transport_layer.udp.send_buffer, eCAL_Config_GetUdpMulticastSndBufSizeBytes());
    EXPECT_EQ(configuration0->transport_layer.udp.send_buffer, eCAL_GetConfiguration()->transport_layer.udp.send_buffer);
    EXPECT_EQ(configuration0->transport_layer.udp.receive_buffer, eCAL_Config_GetUdpMulticastRcvBufSizeBytes());
    EXPECT_EQ(configuration0->transport_layer.udp.receive_buffer, eCAL_GetConfiguration()->transport_layer.udp.receive_buffer);
    EXPECT_EQ(configuration0->transport_layer.udp.join_all_interfaces, eCAL_Config_IsUdpMulticastJoinAllIfEnabled());
    EXPECT_EQ(configuration0->transport_layer.udp.join_all_interfaces, eCAL_GetConfiguration()->transport_layer.udp.join_all_interfaces);
    EXPECT_EQ(configuration0->transport_layer.udp.npcap_enabled, eCAL_Config_IsNpcapEnabled());
    EXPECT_EQ(configuration0->transport_layer.udp.npcap_enabled, eCAL_GetConfiguration()->transport_layer.udp.npcap_enabled);

    EXPECT_STREQ(configuration0->transport_layer.udp.local.group, eCAL_GetConfiguration()->transport_layer.udp.local.group);
    EXPECT_EQ(configuration0->transport_layer.udp.local.ttl, eCAL_GetConfiguration()->transport_layer.udp.local.ttl);

    EXPECT_EQ(configuration0->transport_layer.tcp.number_executor_reader, eCAL_Config_GetTcpPubsubReaderThreadpoolSize());
    EXPECT_EQ(configuration0->transport_layer.tcp.number_executor_reader, eCAL_GetConfiguration()->transport_layer.tcp.number_executor_reader);
    EXPECT_EQ(configuration0->transport_layer.tcp.number_executor_writer, eCAL_Config_GetTcpPubsubWriterThreadpoolSize());
    EXPECT_EQ(configuration0->transport_layer.tcp.number_executor_writer, eCAL_GetConfiguration()->transport_layer.tcp.number_executor_writer);
    EXPECT_EQ(configuration0->transport_layer.tcp.max_reconnections, eCAL_Config_GetTcpPubsubMaxReconnectionAttemps());
    EXPECT_EQ(configuration0->transport_layer.tcp.max_reconnections, eCAL_GetConfiguration()->transport_layer.tcp.max_reconnections);
}

TEST_F(config_test_c, Subscriber)
{
    EXPECT_EQ(configuration0->subscriber.layer.udp.enable, eCAL_Config_IsUdpMulticastRecEnabled());
    EXPECT_EQ(configuration0->subscriber.layer.udp.enable, eCAL_GetConfiguration()->subscriber.layer.udp.enable);
    EXPECT_EQ(configuration0->subscriber.layer.shm.enable, eCAL_Config_IsShmRecEnabled());
    EXPECT_EQ(configuration0->subscriber.layer.shm.enable, eCAL_GetConfiguration()->subscriber.layer.shm.enable);
    EXPECT_EQ(configuration0->subscriber.layer.tcp.enable, eCAL_Config_IsTcpRecEnabled());
    EXPECT_EQ(configuration0->subscriber.layer.tcp.enable, eCAL_GetConfiguration()->subscriber.layer.tcp.enable);
    EXPECT_EQ(configuration0->subscriber.drop_out_of_order_messages, eCAL_Config_GetDropOutOfOrderMessages());
    EXPECT_EQ(configuration0->subscriber.drop_out_of_order_messages, eCAL_GetConfiguration()->subscriber.drop_out_of_order_messages);
}

TEST_F(config_test_c, Time)
{
    EXPECT_STREQ(configuration0->timesync.timesync_module_rt, eCAL_Config_GetTimesyncModuleName());
    EXPECT_STREQ(configuration0->timesync.timesync_module_rt, eCAL_GetConfiguration()->timesync.timesync_module_rt);
    EXPECT_STREQ(configuration0->timesync.timesync_module_replay, eCAL_Config_GetTimesyncModuleReplay());
    EXPECT_STREQ(configuration0->timesync.timesync_module_replay, eCAL_GetConfiguration()->timesync.timesync_module_replay);
}

TEST_F(config_test_c, Application)
{
    EXPECT_STREQ(configuration0->application.startup.terminal_emulator, eCAL_Config_GetTerminalEmulatorCommand());
    EXPECT_STREQ(configuration0->application.startup.terminal_emulator, eCAL_GetConfiguration()->application.startup.terminal_emulator);
    EXPECT_STREQ(configuration0->application.sys.filter_excl, eCAL_Config_GetEcalSysFilterExcludeList());
    EXPECT_STREQ(configuration0->application.sys.filter_excl, eCAL_GetConfiguration()->application.sys.filter_excl);
}

TEST_F(config_test_c, Logging)
{
    EXPECT_EQ(configuration0->logging.provider.console.enable, eCAL_GetConfiguration()->logging.provider.console.enable);
    EXPECT_EQ(configuration0->logging.provider.console.log_level, eCAL_GetConfiguration()->logging.provider.console.log_level);
    EXPECT_EQ(configuration0->logging.provider.file.enable, eCAL_GetConfiguration()->logging.provider.file.enable);
    EXPECT_EQ(configuration0->logging.provider.file.log_level, eCAL_GetConfiguration()->logging.provider.file.log_level);
    EXPECT_STREQ(configuration0->logging.provider.file_config.path, eCAL_GetConfiguration()->logging.provider.file_config.path);
    EXPECT_EQ(configuration0->logging.provider.udp.enable, eCAL_GetConfiguration()->logging.provider.udp.enable);
    EXPECT_EQ(configuration0->logging.provider.udp.log_level, eCAL_GetConfiguration()->logging.provider.udp.log_level);
    EXPECT_EQ(configuration0->logging.provider.udp_config.port, eCAL_GetConfiguration()->logging.provider.udp_config.port);
    EXPECT_EQ(configuration0->logging.receiver.enable, eCAL_GetConfiguration()->logging.receiver.enable);
    EXPECT_EQ(configuration0->logging.receiver.udp_config.port, eCAL_GetConfiguration()->logging.receiver.udp_config.port);
}

TEST_F(config_test_c, Publisher)
{
    EXPECT_EQ(configuration0->publisher.layer.shm.acknowledge_timeout_ms, eCAL_GetConfiguration()->publisher.layer.shm.acknowledge_timeout_ms);
    EXPECT_EQ(configuration0->publisher.layer.shm.enable, eCAL_GetConfiguration()->publisher.layer.shm.enable);
    EXPECT_EQ(configuration0->publisher.layer.shm.memfile_buffer_count, eCAL_GetConfiguration()->publisher.layer.shm.memfile_buffer_count);
    EXPECT_EQ(configuration0->publisher.layer.shm.memfile_min_size_bytes, eCAL_GetConfiguration()->publisher.layer.shm.memfile_min_size_bytes);
    EXPECT_EQ(configuration0->publisher.layer.shm.memfile_reserve_percent, eCAL_GetConfiguration()->publisher.layer.shm.memfile_reserve_percent);
    EXPECT_EQ(configuration0->publisher.layer.shm.zero_copy_mode, eCAL_GetConfiguration()->publisher.layer.shm.zero_copy_mode);
    EXPECT_EQ(configuration0->publisher.layer.tcp.enable, eCAL_GetConfiguration()->publisher.layer.tcp.enable);
    EXPECT_EQ(configuration0->publisher.layer.udp.enable, eCAL_GetConfiguration()->publisher.layer.udp.enable);

    EXPECT_EQ(configuration0->publisher.layer_priority_local_length, eCAL_GetConfiguration()->publisher.layer_priority_local_length);
    EXPECT_EQ(configuration0->publisher.layer_priority_remote_length, eCAL_GetConfiguration()->publisher.layer_priority_remote_length);

    for (int i = 0; i < int(configuration0->publisher.layer_priority_local_length); i++)
    {
        EXPECT_EQ(configuration0->publisher.layer_priority_local[i], eCAL_GetConfiguration()->publisher.layer_priority_local[i]);
    }

    for (int i = 0; i < int(configuration0->publisher.layer_priority_remote_length); i++)
    {
        EXPECT_EQ(configuration0->publisher.layer_priority_remote[i], eCAL_GetConfiguration()->publisher.layer_priority_remote[i]);
    }
}